torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -783,6 +796,36 @@ uint64_t ProcessGroupNCCL::getSequenceNumberForGroup() {
   return seq_;
 }
 
+void ProcessGroupNCCL::registerOnCompletionHook(
+    std::function<void(c10::intrusive_ptr<Work>)>&& hook) {
+  TORCH_CHECK(
+      enableTiming_,
+      "ProcessGroupNCCL OnCompletion hook requires recording start and end "
+      "events which require setting NCCL_ENABLE_TIMING environment variable. "
+      "This is only available for NCCL version >= 2.4.");
+  onCompletionHook_ = std::move(hook);
```

Comment:
This is a fragile design, can't enable recording at this point?

Reply:
ah, yep, good point! Will keep just one hook in this PR and add a check to prevent re-registering. Not sure if we should allow multiple hooks in the future. We can revisit if necesssary.

Reply:
I believe that having an env var gate an optional feature is fragile. It will lead to a lot of extra work from our users when they get the exception.

Why can't we enable recording those events when the callback is set and not have this env var at all?

Reply:
> Why can't we enable recording those events when the callback is set and not have this env var at all?

The original reason what that users might not need enableTiming if they don't need CUDA elapsed time. But I guess that can also be done with a separate argument when register the hook.

Let me make the change to avoid the env var.

Reply:
> Why can't we enable recording those events when the callback is set and not have this env var at all?

Hey @kumpera,  2nd thought on this one. As users might register the hook in the middle of a job where there could be pending collective work items created with `enableTiming=False` and then toggled to true. I.e., users can get duration on some work but not others, which is confusing. This means we need a flush when register hooks and block future collectives until registration is done. I'd prefer to defer this design decision to future PRs. 

I am also hesitate to turn on recording and timing for all start and end events by default, as changing this default behavior has a large blast radius. 

Another option is to add hooks to the pg options arg for PG initialization. However, this API does not align where with other parts of PyTorch. 

thoughts? :)

Reply:
I agree that recording by default requires being a lot more careful.

Registering hooks in the middle of a job will be confusing regardless. For example, do we fire the hook only for later collectives or is it ok to racy'ish fire for currently running ones?

Reply:
> Registering hooks in the middle of a job will be confusing regardless. For example, do we fire the hook only for later collectives or is it ok to racy'ish fire for currently running ones?

Not sure yet. I'd vote for we should fire hooks only for later collectives. Or, maybe we should ban this, where hooks can only be registered before firing any collectives, otherwise error out?